pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -0,0 +1,199 @@
+						triggerIdsForReports[triggerId] = make([]int, 0)
+					}
+					triggerIdsForReports[triggerId] = append(triggerIdsForReports[triggerId], reportIndex)
+				}
+			}
+		}
+		reportIndex += 1
+	}
+
+	// Then for each trigger id, find which reports correspond to that trigger and create an event bundling the reports
+	// and send it to the channel associated with the trigger id.
+	for triggerId, reportIds := range triggerIdsForReports {
+		reportPayload := make([]MercuryReport, 0)
+		for _, reportId := range reportIds {
+			reportPayload = append(reportPayload, reports[reportId])
+		}
+
+		triggerEvent := MercuryTriggerEvent{
+			triggerType: "mercury",
+			id:          generateTriggerEventId(reportPayload),
+			timestamp:   strconv.FormatInt(unixTimestampMillis, 10),
+			payload:     reportPayload,
+		}
+
+		// TODO: Modify values.Wrap to handle MercuryTriggerEvent and MercuryReport structs
+		val, err := values.Wrap(triggerEvent)
+		if err != nil {
+			return err
+		}
+
+		// Create a new CapabilityResponse with the MercuryTriggerEvent
+		capabilityResponse := capabilities.CapabilityResponse{
+			Value: val,
+		}
+
+		// If the FeedId is in the feedIdsForTriggerId map, then send the event to the channel.
+
+		wfID := o.triggerIdToWorkflowId[triggerId]
+		ch, ok := o.chans[workflowID(wfID)]
+		if ok {
+			ch <- capabilityResponse
+		} else {
+			return fmt.Errorf("no registration for %s", triggerId)
+		}
+	}
+	return nil
+}
+
+func (o *MercuryTriggerService) RegisterTrigger(ctx context.Context, callback chan<- capabilities.CapabilityResponse, req capabilities.CapabilityRequest) error {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.chans[workflowID(wid)] = callback // TODO: if this workflowID is already in the map, throw an error?
+	// set feedIdsForTriggerId
+	triggerId := o.GetTriggerId(req) // TODO: This is manadatory, so throw an error if its not present or already exists
+	feedIds := o.GetFeedIds(req)     // TODO: what if feedIds is empty? should we throw an error or allow it?
+	o.feedIdsForTriggerId[triggerId] = feedIds
+
+	o.triggerIdToWorkflowId[triggerId] = wid
+	return nil
+}
+
+func (o *MercuryTriggerService) UnregisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) error {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	ch, ok := o.chans[workflowID(wid)]
+	if ok {
+		close(ch)
+	}
+
+	triggerId := o.GetTriggerId(req)
+
+	delete(o.chans, workflowID(wid))
+	delete(o.feedIdsForTriggerId, triggerId)
+	delete(o.triggerIdToWorkflowId, triggerId)
+
+	return nil
+}
+
+// Get array of feedIds from CapabilityRequest req
+func (o *MercuryTriggerService) GetFeedIds(req capabilities.CapabilityRequest) []uint64 {
+	feedIds := make([]uint64, 0)
+	// Unwrap the inputs which should return pair (map, nil) and then get the feedIds from the map
+	if inputs, err := req.Inputs.Unwrap(); err == nil {
+		if feeds, ok := inputs.(map[string]interface{})["feedIds"].([]uint64); ok {
```

Comment:
According to Austin, in current Mercury version we should use [32]byte as feed IDs (or just a byte array and we can validate len separately).

Reply:
@bolekk why [32]byte rather than uint64? I know @samsondav is using the latter for LLO

Reply:
I will handle this in a separate PR.